bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -380,6 +380,16 @@ void ReplicaImp::onMessage<ClientRequestMsg>(ClientRequestMsg *m) {
       return;
     }
   }
+  if ((flags & KEY_EXCHANGE_FLAG) && repsInfo->isIdOfPeerReplica(senderId)) {
```

Comment:
You can have this logic in the onKeyExchange inside the keyExchangeManager.
Here https://github.com/vmware/concord-bft/blob/master/bftengine/src/bftengine/KeyExchangeManager.cpp#L68 the `sn` argument is for the bft sequence number of the prePrepare containing this keyExchange message.
In addition, you have the generated sequence number in the message itself.
I think it would be more clear to have this logic there